src/hotspot/share/runtime/stackOverflow.hpp

```diff
@@ -120,6 +132,81 @@ class StackOverflow {
   //
   //  (high addresses)
   //
+  //
+  // The stack overflow mechanism detects overflows by touching ("banging") the stack
+  // ahead of current stack pointer (SP). The entirety of guard zone is memory protected,
+  // therefore such access would trap when touching the guard zone, and one of the following
+  // things would happen.
+  //
+  // Access in the red zone: unrecoverable stack overflow. Crash the VM, generate a report,
+  // crash dump, and other diagnostics.
+  //
+  // Access in the yellow zone: recoverable, reportable stack overflow. Create and throw
+  // a StackOverflowError, remove the protection of yellow zone temporarily to let exception
+  // handlers run. If exception handlers themselves run out of stack, they will crash VM due
+  // to access to red zone.
+  //
+  // Access in the reserved zone: recoverable, reportable, transparent for privileged methods
+  // stack overflow. Perform a stack walk to check if there's a method annotated with
+  // @ReservedStackAccess on the call stack. If such method is found, remove the protection of
+  // reserved zone temporarily, and let the method run. If not, handle the access like a yellow
+  // zone trap.
+  //
+  // The banging itself happens within the "shadow zone" that extends from the current SP.
+  //
+  // The goals for properly implemented shadow zone banging are:
+  //
+  //  a) Allow native/VM methods to run without stack overflow checks within some reasonable
+  //     headroom. Default shadow zone size should accommodate the largest normally expected
+  //     native/VM stack use.
+  //  b) Guarantee the stack overflow checks work even if SP is dangerously close to guard zone.
+  //     If SP is very low, banging at the edge of shadow zone (SP+shadow-zone-size) can slip
+  //     into adjacent thread stack, or even into other readable memory. This would potentially
+  //     pass the check by accident.
+  //  c) Allow for incremental stack growth by handling traps from not yet committed thread
```

Comment:
I failed to understand why we have to do "incremental stack growth" here.  Why can't use touch the last page? 
```
__ bang_stack_with_offset(n_shadow_pages*page_size);
```

The entire shadow zone is mapped. Touching it causes commit, page faults or SEGV. First 2 events are transparent for the userspace process.

Hotspot will trap into the signal handler if `bang_stack_shadow_pages` does cross shadow_zone_safe_limit().  `rsp + n_shadow_pages * page_size` falls into 2 possibilities: 
1. red zone:  the program is about to die anyway. 
2. yellow reserved zones, both are recoverable.

I feel it's not necessary to touch pages from 1 to n_shadow_pages-1. The side effect is same as touching the last page directly.

ps: I tried this [idea](https://github.com/navyxliu/jdk/runs/5075962312?check_suite_focus=true). 2 failures are found on Windows. I guess the premise that the shadow zone is mapped is false on Windows. 
```
compiler/interpreter/cr7116216/StackOverflow.java 
compiler/uncommontrap/UncommonTrapStackBang.java 
```

Reply:
I read this blogpost and I need to take back my comment. 
https://pangin.pro/posts/stack-overflow-handling

now I think interpreter has to do linear probing to make sure HotSpot executes Java programs correctly. reserve_zone has special meaning.  Further, if rsp is very close to the shadow_zone_safe_limit(), so-called last page may surpass red zone.

Reply:
Yes, that's exactly what point "(c)" is about.